pkg/apis/openshiftapiserver/v1alpha1/types.go

```diff
@@ -21,24 +21,28 @@ type OpenShiftAPIServerConfig struct {
 // OpenShiftAPIServerOperatorConfig provides information to configure an operator to manage openshift-apiserver.
 type OpenShiftAPIServerOperatorConfig struct {
 	metav1.TypeMeta   `json:",inline"`
-	metav1.ObjectMeta `json:"metadata" protobuf:"bytes,1,opt,name=metadata"`
+	metav1.ObjectMeta `json:"metadata"`
 
-	Spec   OpenShiftAPIServerOperatorConfigSpec   `json:"spec" protobuf:"bytes,2,opt,name=spec"`
-	Status OpenShiftAPIServerOperatorConfigStatus `json:"status" protobuf:"bytes,3,opt,name=status"`
+	Spec   OpenShiftAPIServerOperatorConfigSpec   `json:"spec"`
+	Status OpenShiftAPIServerOperatorConfigStatus `json:"status"`
 }
 
 type OpenShiftAPIServerOperatorConfigSpec struct {
-	operatorsv1alpha1api.OperatorSpec `json:",inline" protobuf:"bytes,1,opt,name=operatorSpec"`
+	operatorsv1alpha1api.OperatorSpec `json:",inline"`
 
-	// OpenShiftAPIServerConfig holds a sparse config that the user wants for this component.  It only needs to be the overrides from the defaults
+	// userConfig holds a sparse config that the user wants for this component.  It only needs to be the overrides from the defaults
 	// it will end up overlaying in the following order:
 	// 1. hardcoded default
 	// 2. this config
-	OpenShiftAPIServerConfig runtime.RawExtension `json:"kubeApiserverConfig" protobuf:"bytes,2,opt,name=kubeApiserverConfig"`
+	UserConfig runtime.RawExtension `json:"userConfig"`
+
+	// observedConfig holds a sparse config that controller has observed from the cluster state.  It exists in spec because
+	// it causes action for the operator
+	ObservedConfig runtime.RawExtension `json:"observedConfig"`
```

Comment:
@smarterclayton this field

Reply:
The `observedConfig` is an input to the level that determines an overall config.  The overall config (logically the merging of default, user, and observed) is a level (reflected by spec.generation) that the operator will try to reconcile.  Once it has successfully reconciled the level, status.generation will indicate that the level has been achieved.  Since it is an input the level, it should live in the spec, not the status.

Reply:
> If we decide this was a terrible idea in the future, can we change it? I.e. can we classify these APIs in a way that gives us more flexibility than we might traditionally have to change their internal workings?

Yes.  Since they will be immediately stomped by a controller, the "nothing you ever set here had an effect" rule applies to changes.

Reply:
API wise, to me that name sounds ambiguous - in the controller context *observed* means something that the controller saw and acted upon. I don't have a good fitting name it though.

> I can see the reasoning for this, and it's probably the best way to manage complex config opaquely.

@smarterclayton 

Is there a reason why we need to cache the value? Currently it are the etcd urls - getting them from endpoints informer is as cheep as it gets. Caching the field on the object itself leads to having 2 controllers on the same object just to fill in this field, introducing complexity, loosing atomicity, RV collisions, likely broken first rollout until the other loop catches up, ...

I would imagine this config detected from cluster state is determined in a regular sync loop from informer caches, merged with what we have in the spec and not stored on the operator config object.

I disagree with the generation implications presented here. Generation is good for letting user know which version the controller saw and acted upon. We shouldn't trigger based on generation but on actual!=desired. Like when in sync loop you merge UserConfig and etcd values you read from endpoints and the result doesn't match what the running apiserver is configured with. If we don't need it for triggering, as generation is not what we should be triggering on, and it is not something anyone should set and we don't need to cache it, I don't see a reason to have this field and multiple controllers coupled to it.

/cc @mfojtik